Add a SequenceLearner

basnijholt opened the pull request:
Adaptive has really great features to trivially parallelize and auto-save your simulation, however, sometimes you just can't formulate your problem in one of the `Learner`s. Then just for this single problem, one needs to write a lot of code to basically do what Adaptive can already do for you.

Therefore I've implemented a `SequenceLearner` that doesn't do a whole lot of fancy things, except for being a `BaseLearner` child.

With this one can easily run, `save` (and periodically save), `load`, `live_info`, and `live_plot` one's progress (if one implements a simple `plotter` function).

It's great when running simulations that might crash because if one uses periodic saving, the learner just picks up where it was before it crashed (in contrary to `ipyparallel.Client().map` where one needs to implement all those features).

@jbweston and @akhmerov, what do you guys think? If this doesn't belong in Adaptive I'll put it in another package.


Example:
```python
def f(x):
    x = int(x)
    return x ** 2

seq = range(1000)
learner = SequenceLearner(f, seq)
adaptive.runner.simple(learner, SequenceLearner.done)
# that goal is same as `lambda l: l.done()`
```

```python
def plotter(learner):
    data = learner.data if learner.data else []
    return hv.Scatter(data)

learner = SequenceLearner(f, seq)
runner = adaptive.Runner(learner, SequenceLearner.done)
runner.live_plot(plotter=plotter)
```

```python
def g(xy):
    x, y = xy
    return x ** 2 + y**2

seq = np.random.rand(1000, 2)
learner = SequenceLearner(g, seq)
runner = adaptive.Runner(learner, SequenceLearner.done)
runner.live_info()
```

basnijholt commented:
I did some testing to see how Adaptive compares with `ipyparallel`:
```
import adaptive
adaptive.notebook_extension()
import asyncio
import ipyparallel
c = ipyparallel.Client()
lview = c.load_balanced_view()
```

```
def f(x, wait=False):
    if wait:
        import time
        time.sleep(1)
    return 1

seq = list(range(5000))
%time res = lview.map_sync(f, seq)
```

Using a local `ipcluster` with 48 local engines:
- `wait=False` 2.77 s (`lview`) vs 2.82 s (`adaptive.Runner`) 
- `wait=True` 22.5 (`lview`) vs 22.8 s (`adaptive.Runner`) 

On the cluster, 100 engines on the `hpc05` (cluster) and the notebook on `io` (different machine outside of the cluster):
- `wait=False` 13.8 s (`lview`) vs 16.4 s (`adaptive.Runner`) 
- `wait=True` 55 s (`lview`) vs 51 s (`adaptive.Runner`) 

Surprisingly. somehow Adaptive is faster in the last case!

basnijholt commented:
Until this is merged the `SequenceLearner` also lives here https://github.com/basnijholt/adaptive-scheduler/blob/master/adaptive_scheduler/sequence_learner.py.

The requirement for the sequence to be hashable might be too strict. We could instead make `self._to_do_seq` into `self._to_do_indices` and pass the index to the function which would then internally be wrapped with `functools.partial` and call the function with the right parameters, similar to what we do in the `BalancingLearner`.

akhmerov commented:
While I don't think this is a useful addition to the library, it also does not extend the API and shouldn't break anything (if implemented correctly). Therefore it may be a useful example to those wishing to implement their own learner (or those, indeed, who find the non-adaptive part of adaptive relevant).

However in order there are several changes that I would like to suggest, that would make a conceptual sense:

- Currently the points are not requested in order (sets aren't sorted), we should use a sorted container.
- The loss should be equal to `len(_to_do_seq) / len(sequence)`
- As a corollary, the loss improvement should always be equal to `1/len(sequence)`
- If the original points are numpy arrays, the learner will be requesting tuples, and potentially breaking user-provided functions (e.g. if they call an array method on the point). We should be careful to not change the type of the point.
- We should add a test that the learner correctly requests points in the sequence that was provided by the user.
- The docstring should explain what the learner actually does.

akhmerov commented:
Actually even easier: loss improvement per point should be just `1`.

basnijholt commented:
@akhmerov I've implemented your suggestions.

In order to use and pass the exact elements in the sequence that the user passes, I've changed `_to_do_seq` into `_to_do_indices`.

Because `self.pending_points` and self.data` depend on something that is hashable (and the user's input might not be hashable) we use the indices of the input. This means that the user's function receives a tuple `(index, point)` and needs to be wrapped in order to ignore the index.

jbweston commented on adaptive/learner/sequence_learner.py:
```diff
@@ -0,0 +1,132 @@
+
+    This is useful when your problem cannot be formulated in terms of
+    another adaptive learner, but you still want to use Adaptive's
+    routines to run, save, and plot.
```
I feel there could be a little more explanation about what the learner does. I'm not sure if "learning a sequence" is very clearn.

jbweston replied:
something that explains that the learner simply gives back the points in the provided sequence when `ask`ed.

jbweston commented on adaptive/learner/sequence_learner.py:
```diff
@@ -0,0 +1,132 @@
+
+from adaptive.learner.base_learner import BaseLearner
+
+inf = sys.float_info.max
```
This isn't needed anymore
